presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/JdbcMetadata.java

```diff
@@ -114,7 +114,7 @@ public JdbcTableHandle getTableHandle(ConnectorSession session, SchemaTableName
         }
 
         if (handle.getLimit().isPresent() && handle.getLimit().getAsLong() <= limit) {
-            return Optional.empty();
+            return Optional.of(new LimitApplicationResult<>(handle, jdbcClient.isLimitGuaranteed()));
```

Comment:
Isn't this going to loop now? You can apply same limit over and over again.

Reply:
My idea was that I'm _swallowing_ a limit.
But indeed, I can do this only when it's guaranteed.

Tests seem to agree with you:
```
[ERROR] testLimitPushDown(io.prestosql.plugin.mysql.TestMySqlDistributedQueries)  Time elapsed: 0.689 s  <<< FAILURE!
java.lang.RuntimeException: The optimizer exhausted the time limit of 180000 ms
```